# nf-basemap

[Nextflow] workflow for basecalling and mapping of whole genome or whole transcriptome Nanopore fastq reads using dorado, minimap2, samtools and multiqc.

## Dependencies

Requires either [Docker] or [Apptainer] installed.

Usage:

```sh
nextflow run chusj-pigu/nf-basemap -r main [--pod5 INPUT] | [--fastq INPUT] [--ref REF_GENOME] [OPTIONS]
```

To list available options and parameters for this workflow, run :

``` sh
nextflow run chusj-pigu/nf-basemap -r main --help
```

Overview:

This workflow can be run locally or on Compute Canada. To use on Compute Canada, use the `-profile drac` option. For testing purposes, basecalling can be run on cpu only with `-profile test`.

### Parameters

- `--pod5`: Path to the directory containing pod5.
- `--fastq`: Path to the directory containing fastq files. Only use if skipping basecalling.
- `--ref`: Path to the reference fasta file for alignment.
- `--skip_basecall`: Basecalling step will be skipped; input must be in fastq [default: false].
- `--skip_mapping`: Mapping will be skipped [default: false].
- `--duplex`: Dorado will basecall in duplex mode instead of simplex [default: false].
- `--out_dir`: Output directory to place mapped files and reports in [default: output].
- `--sample_id`: Will name output files according to sample id [default: reads].
- `--no_mod`: Basecalling without base modification [default: false].
- `-profile`: Use standard for running locally, or drac when running on Digital Research Alliance of Canada Narval [default: standard].
- `--bed`: Bed file containing regions of interest to compute mapping statistics with mosdepth .
- `--batch`: Batch size for basecalling; if 0, optimal batch size will be automatically selected [default: 0].
When running locally:
- `--model`: Basecalling model to use [default: sup@v5.0.0].
- `--m_bases`: Modified bases to be called, separated by commas if more than one is desired. Requires path to model if run with drac profile [default: 5mC_5hmC].
When running on Narval:
- `--model_path`: Path for the basecalling model, required when running with drac profile [default: path to sup@v5.0.0].
- `--m_bases_path`: Path for the modified basecalling model, required when running with drac profile [default: path to sup@v5.0.0_5mC_5hmC].

## 1. Basecalling

Basecalling is done with [Dorado] in simplex mode with modified base calling by default, using the Super accurate algorithm. The pod5 files are first split by channel to facilitate duplex calling. To use without modified basecalling, use the `--no_mod` option. This can be done in duplex mode using the `--duplex` option, or entirely skipped using the `--skip_basecall` option.

## 2. QC stats and filtering

QC plots are generated by [NanoPlot]. Reads that have a QS score <10 will be classified as failed, and subsequent steps are only carried out on reads with QS score >= 10.

## 3. Mapping

If using `--skip_basecall`, the workflow will start here using `--fastq` as input. Mapping of the reads is done with [minimap2] using `-ax -map-ont` options for long reads generated by Nanopore sequencing. Mapping can be skipped with `--skip_mapping`.

## 4. Sorting and indexing

Using [samtools], the .sam file is first converted to .bam using `samtools view`, sorted with `samtools sort` and indexed using `samtools index`. Mapping statistics are computed with [mosdepth].

## Outputs

Generates a .sam file, .bam files for sorted reads and bai for indexed reads. Mapping statistics are stored in text files and summarized in a [multiqc] report.

[Docker]: https://www.docker.com
[Apptainer]: https://apptainer.org
[Nextflow]: https://www.nextflow.io/docs/latest/index.html
[Dorado]: https://github.com/nanoporetech/dorado
[minimap2]: https://lh3.github.io/minimap2/minimap2.html
[samtools]: http://www.htslib.org
[multiqc]: https://multiqc.info
[mosdepth]: https://github.com/brentp/mosdepth
[NanoPlot]: https://github.com/wdecoster/NanoPlot
